docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,287 @@
+  
+- add new page into `f.pendingPages`
+  - increase `f.pendingReaders`
+  - return `f.pendingChan`
+* Range: 
+
+  `Range` returns a slice of mmaped file (`f.data[start:end]`). 
+
+  It will not try to fetch data from obj store, and return an error if some pages in the range have not been fetched.
+
+  `Range` internals:
+
+  - convert range to page ids 
+  - if all page ids have been fetched
+    - return a slice of mmaped file
+
+  * otherwise return an error
+
+* CloseMmap: `CloseMmap` unmmaps `f.data` .
+
+* CloseFile: `CloseFile` closes `f.file`. 
+
+* Close: `Close` closes `f.file`, unmmaps `f.data` and set `f.pages` to nil.
+
+### Range to pages
+
+Convert a range to pages:
+
+```
+const pageSize = 4 * 1024  // 4KiB
+
+minPageID := math.Floor( float(start) / float(pageSize) )
+maxPageID := math.Ceil( float(end) / float(pageSize) )
+
+pages := roaring.NewBitmap()
+pages.AddRange(minPageID, maxPageID)
+```
+
+### Combine queries
+
+```
+type BucketFile struct {
+    // ...
+    pendingDuration time.Duration   // duration to fetch pending pages
+    pendingPages    *roaring.Bitmap // pending pages to be fetched
+    pendingReaders  int             // record pending callers
+    pendingChan     chan error      // chan to notify callers the result of prefetch	
+}
+
+func (f *BucketFile) Prefetch(ranges []*Range) (chan error, error) {...}
+```
+
+BucketFile appends new pages to `pendingPages`, waits for next time to fetch. So that we can combine multiple queries to ones.
+
+If the combined prefetch fails, all sub prefetches will fail. There must be error handling for failed prefetches.
+
+Each time to fetch pending pages, BucketFile will reset `pendingPages`, `pendingReaders` and `pendingChan`.
+
+### In-memory objects
+
+Each IndexReader decodes [postings offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) into in-memory object for sequential and random accesses.
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses small memory. 
+
+For inactive IndexReaders and BucketFiles, if the inactive duration (`now.Sub(lastQueryTime)`) is greater than a configured duration (e.g. 15 mins), then IndexReaders will remove in-memory objects, and BucketFiles will remove in-memory bitmaps, close file descriptors and unmmap local files. This will reduce memory usage.
+
+### Preload blocks on startup
+
+The latest blocks will be queried the most. 
+
+Preload `the last N hours` (configurable) blocks on startup.
+This will fetch index files from object storage and decode bytes into in-memory objects.
+
+### Local files
+
+Because the max size of a index file is 64 GiB, and the default size of a chunks file is 512 MiB.
+
+So we can set the size of a local index file to 64 GiB, set the size of a local chunks file to 768 (512*1.5) MiB.
```

Comment:
Do we really need to know/set max size of files here? Why?

Reply:
`mmap` needs a size to allocate virtual address space. If we give a small size, when the offset of fetched data is greater than mmap size,  we need to re-mmap.